File: genkit-tools/common/src/eval/localFileDatasetStore.ts
/**
 * Copyright 2024 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import fs from 'fs';
import { readFile, rm, writeFile } from 'fs/promises';
import path from 'path';
import { v4 as uuidv4 } from 'uuid';
import type { CreateDatasetRequest, UpdateDatasetRequest } from '../types/apis';
import {
  DatasetSchema,
  type Dataset,
  type DatasetMetadata,
  type DatasetStore,
  type InferenceDataset,
} from '../types/eval';
import { generateTestCaseId } from '../utils';
import { logger } from '../utils/logger';

/**
 * A local, file-based DatasetStore implementation.
 */
export class LocalFileDatasetStore implements DatasetStore {
  private readonly storeRoot;
  private readonly indexFile;
  private static cachedDatasetStore: LocalFileDatasetStore | null = null;

  private constructor(storeRoot: string) {
    this.storeRoot = storeRoot;
    this.indexFile = this.getIndexFilePath();
    if (!fs.existsSync(this.storeRoot)) {
      fs.mkdirSync(this.storeRoot, { recursive: true });
    }
    if (!fs.existsSync(this.indexFile)) {
      fs.writeFileSync(path.resolve(this.indexFile), JSON.stringify({}));
    }
    logger.debug(
      `Initialized local file dataset store at root: ${this.storeRoot}`
    );
  }

  static getDatasetStore() {
    if (!this.cachedDatasetStore) {
      this.cachedDatasetStore = new LocalFileDatasetStore(
        this.generateRootPath()
      );
    }
    return this.cachedDatasetStore;
  }

  static reset() {
    this.cachedDatasetStore = null;
  }

  async createDataset(req: CreateDatasetRequest): Promise<DatasetMetadata> {
    return this.createDatasetInternal(req);
  }

  private async createDatasetInternal(
    req: CreateDatasetRequest
  ): Promise<DatasetMetadata> {
    const { data, datasetId, schema, targetAction } = req;
    const id = await this.generateDatasetId(datasetId);
    const filePath = path.resolve(this.storeRoot, this.generateFileName(id));

    if (fs.existsSync(filePath)) {
      logger.error(`Dataset already exists at ` + filePath);
      throw new Error(
        `Create dataset failed: file already exists at {$filePath}`
      );
    }
    const dataset = this.getDatasetFromInferenceDataset(data);
    logger.debug(`Saving Dataset to ` + filePath);
    await writeFile(filePath, JSON.stringify(dataset));

    const now = new Date().toString();
    const metadata: DatasetMetadata = {
      datasetId: id,
      schema,
      targetAction,
      metricRefs: req.metricRefs,
      size: dataset.length,
      version: 1,
      datasetType: req.datasetType,
      createTime: now,
      updateTime: now,
    };

    const metadataMap = await this.getMetadataMap();
    metadataMap[id] = metadata;

    logger.debug(
      `Saving DatasetMetadata for ID ${id} to ` + path.resolve(this.indexFile)
    );

    await writeFile(path.resolve(this.indexFile), JSON.stringify(metadataMap));
    return metadata;
  }

  async updateDataset(req: UpdateDatasetRequest): Promise<DatasetMetadata> {
    const { datasetId, data, schema, targetAction, metricRefs } = req;
    const filePath = path.resolve(
      this.storeRoot,
      this.generateFileName(datasetId)
    );
    if (!fs.existsSync(filePath)) {
      throw new Error(`Update dataset failed: dataset not found`);
    }

    const metadataMap = await this.getMetadataMap();
    const prevMetadata = metadataMap[datasetId];
    if (!prevMetadata) {
      throw new Error(`Update dataset failed: dataset metadata not found`);
    }
    const patch = this.getDatasetFromInferenceDataset(data ?? []);
    let newSize = prevMetadata.size;
    if (patch.length > 0) {
      logger.debug(`Updating Dataset at ` + filePath);
      newSize = await this.patchDataset(datasetId, patch, filePath);
    }

    const now = new Date().toString();
    const newMetadata = {
      datasetId: datasetId,
      size: newSize,
      schema: schema ? schema : prevMetadata.schema,
      targetAction: targetAction ? targetAction : prevMetadata.targetAction,
      version: data ? prevMetadata.version + 1 : prevMetadata.version,
      datasetType: prevMetadata.datasetType,
      metricRefs: metricRefs ? metricRefs : prevMetadata.metricRefs,
      createTime: prevMetadata.createTime,
      updateTime: now,
    };

    logger.debug(
      `Updating DatasetMetadata for ID ${datasetId} at ` +
        path.resolve(this.indexFile)
    );
    // Replace the metadata object in the metadata map
    metadataMap[datasetId] = newMetadata;
    await writeFile(path.resolve(this.indexFile), JSON.stringify(metadataMap));

    return newMetadata;
  }

  async getDataset(datasetId: string): Promise<Dataset> {
    const filePath = path.resolve(
      this.storeRoot,
      this.generateFileName(datasetId)
    );
    if (!fs.existsSync(filePath)) {
      throw new Error(`Dataset not found for dataset ID ${datasetId}`);
    }
    return await readFile(filePath, 'utf8').then((data) => {
      return DatasetSchema.parse(JSON.parse(data));
    });
  }

  async listDatasets(): Promise<DatasetMetadata[]> {
    return this.getMetadataMap().then((metadataMap) => {
      const metadatas = [];

      for (var key in metadataMap) {
        metadatas.push(metadataMap[key]);
      }
      return metadatas;
    });
  }

  async deleteDataset(datasetId: string): Promise<void> {
    const filePath = path.resolve(
      this.storeRoot,
      this.generateFileName(datasetId)
    );
    await rm(filePath);

    const metadataMap = await this.getMetadataMap();
    delete metadataMap[datasetId];

    logger.debug(
      `Deleting DatasetMetadata for ID ${datasetId} in ` +
        path.resolve(this.indexFile)
    );
    await writeFile(path.resolve(this.indexFile), JSON.stringify(metadataMap));
  }

  private static generateRootPath(): string {
    return path.resolve(process.cwd(), `.genkit/datasets`);
  }

  /** Visible for testing */
  async generateDatasetId(datasetId?: string): Promise<string> {
    const metadataMap = await this.getMetadataMap();
    const keys = Object.keys(metadataMap);
    if (datasetId) {
      const isValid = /^[A-Za-z][A-Za-z0-9_.-]{4,34}[A-Za-z0-9]$/g.test(
        datasetId
      );
      if (!isValid) {
        throw new Error(
          'Invalid datasetId provided. ID must be alphanumeric, with hyphens, dots and dashes. Is must start with an alphabet, end with an alphabet or a number, and must be 6-36 characters long.'
        );
      }
      return this.testUniqueness(datasetId, keys);
    }

    const id = uuidv4();
    return this.testUniqueness(id, keys);
  }

  private testUniqueness(datasetId: string, keys: string[]) {
    if (!keys.some((i) => i === datasetId)) {
      return datasetId;
    }
    throw new Error(`Dataset ID not unique: ${datasetId}`);
  }

  private generateFileName(datasetId: string): string {
    return `${datasetId}.json`;
  }

  private getIndexFilePath(): string {
    return path.resolve(this.storeRoot, 'index.json');
  }

  private async getMetadataMap(): Promise<Record<string, DatasetMetadata>> {
    if (!fs.existsSync(this.indexFile)) {
      return Promise.resolve({} as any);
    }
    return await readFile(path.resolve(this.indexFile), 'utf8').then((data) =>
      JSON.parse(data)
    );
  }

  private getDatasetFromInferenceDataset(data: InferenceDataset): Dataset {
    return data.map((d) => ({
      testCaseId: d.testCaseId ?? generateTestCaseId(),
      ...d,
    }));
  }

  private async patchDataset(
    datasetId: string,
    patch: Dataset,
    filePath: string
  ): Promise<number> {
    const existingDataset = await this.getDataset(datasetId);
    const datasetMap = new Map(existingDataset.map((d) => [d.testCaseId, d]));
    const patchMap = new Map(patch.map((d) => [d.testCaseId, d]));

    patchMap.forEach((value, key) => {
      // Delete sample if testCaseId is provided
      if (value.testCaseId && !value.input && !value.reference) {
        datasetMap.delete(key);
      } else {
        datasetMap.set(key, value);
      }
    });

    const newDataset = Array.from(datasetMap.values()) as Dataset;
    await writeFile(filePath, JSON.stringify(newDataset));
    return newDataset.length;
  }
}

File: java/ai/src/main/java/com/google/genkit/ai/evaluation/LocalFileDatasetStore.java
/*
 * Copyright 2025 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 */

package com.google.genkit.ai.evaluation;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.time.Instant;
import java.util.*;
import java.util.concurrent.ConcurrentHashMap;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import com.fasterxml.jackson.core.type.TypeReference;
import com.google.genkit.core.JsonUtils;

/**
 * File-based implementation of DatasetStore.
 * 
 * <p>
 * Stores datasets in the .genkit/datasets directory with:
 * <ul>
 * <li>index.json - metadata for all datasets</li>
 * <li>{datasetId}.json - actual dataset data</li>
 * </ul>
 */
public class LocalFileDatasetStore implements DatasetStore {

  private static final Logger logger = LoggerFactory.getLogger(LocalFileDatasetStore.class);

  private static final String GENKIT_DIR = ".genkit";
  private static final String DATASETS_DIR = "datasets";
  private static final String INDEX_FILE = "index.json";

  private static LocalFileDatasetStore instance;

  private final Path storeRoot;
  private final Path indexFile;
  private final Map<String, DatasetMetadata> indexCache;

  /**
   * Gets the singleton instance of the dataset store.
   *
   * @return the dataset store instance
   */
  public static synchronized LocalFileDatasetStore getInstance() {
    if (instance == null) {
      instance = new LocalFileDatasetStore();
    }
    return instance;
  }

  /**
   * Creates a new LocalFileDatasetStore using the default location.
   */
  public LocalFileDatasetStore() {
    this(Paths.get(System.getProperty("user.dir"), GENKIT_DIR, DATASETS_DIR));
  }

  /**
   * Creates a new LocalFileDatasetStore with a custom root path.
   *
   * @param storeRoot
   *            the root directory for storing datasets
   */
  public LocalFileDatasetStore(Path storeRoot) {
    this.storeRoot = storeRoot;
    this.indexFile = storeRoot.resolve(INDEX_FILE);
    this.indexCache = new ConcurrentHashMap<>();
    initializeStore();
  }

  private void initializeStore() {
    try {
      Files.createDirectories(storeRoot);
      if (!Files.exists(indexFile)) {
        saveIndex(new HashMap<>());
      }
      loadIndex();
    } catch (IOException e) {
      logger.error("Failed to initialize dataset store", e);
      throw new RuntimeException("Failed to initialize dataset store", e);
    }
  }

  private void loadIndex() throws IOException {
    if (Files.exists(indexFile)) {
      String content = Files.readString(indexFile);
      Map<String, DatasetMetadata> index = JsonUtils.getObjectMapper().readValue(content,
          new TypeReference<Map<String, DatasetMetadata>>() {
          });
      indexCache.clear();
      indexCache.putAll(index);
    }
  }

  private void saveIndex(Map<String, DatasetMetadata> index) throws IOException {
    String json = JsonUtils.toJson(index);
    Files.writeString(indexFile, json);
  }

  private Path getDatasetFile(String datasetId) {
    return storeRoot.resolve(datasetId + ".json");
  }

  private String generateDatasetId() {
    return "dataset_" + UUID.randomUUID().toString().replace("-", "").substring(0, 12);
  }

  @Override
  public DatasetMetadata createDataset(CreateDatasetRequest request) throws Exception {
    String datasetId = request.getDatasetId();
    if (datasetId == null || datasetId.isEmpty()) {
      datasetId = generateDatasetId();
    }

    if (indexCache.containsKey(datasetId)) {
      throw new IllegalArgumentException("Dataset already exists: " + datasetId);
    }

    List<DatasetSample> data = request.getData();
    if (data == null) {
      data = new ArrayList<>();
    }

    // Ensure all samples have testCaseIds
    for (int i = 0; i < data.size(); i++) {
      DatasetSample sample = data.get(i);
      if (sample.getTestCaseId() == null || sample.getTestCaseId().isEmpty()) {
        sample.setTestCaseId("test_case_" + (i + 1));
      }
    }

    String now = Instant.now().toString();
    DatasetMetadata metadata = DatasetMetadata.builder().datasetId(datasetId).size(data.size())
        .schema(request.getSchema())
        .datasetType(request.getDatasetType() != null ? request.getDatasetType() : DatasetType.UNKNOWN)
        .targetAction(request.getTargetAction())
        .metricRefs(request.getMetricRefs() != null ? request.getMetricRefs() : new ArrayList<>()).version(1)
        .createTime(now).updateTime(now).build();

    // Save the dataset data
    String dataJson = JsonUtils.toJson(data);
    Files.writeString(getDatasetFile(datasetId), dataJson);

    // Update the index
    indexCache.put(datasetId, metadata);
    saveIndex(indexCache);

    logger.info("Created dataset: {} with {} samples", datasetId, data.size());
    return metadata;
  }

  @Override
  public DatasetMetadata updateDataset(UpdateDatasetRequest request) throws Exception {
    String datasetId = request.getDatasetId();
    if (datasetId == null || datasetId.isEmpty()) {
      throw new IllegalArgumentException("Dataset ID is required");
    }

    DatasetMetadata existing = indexCache.get(datasetId);
    if (existing == null) {
      throw new IllegalArgumentException("Dataset not found: " + datasetId);
    }

    List<DatasetSample> data = request.getData();
    int size = existing.getSize();

    if (data != null) {
      // Ensure all samples have testCaseIds
      for (int i = 0; i < data.size(); i++) {
        DatasetSample sample = data.get(i);
        if (sample.getTestCaseId() == null || sample.getTestCaseId().isEmpty()) {
          sample.setTestCaseId("test_case_" + (i + 1));
        }
      }
      size = data.size();

      // Save the updated dataset data
      String dataJson = JsonUtils.toJson(data);
      Files.writeString(getDatasetFile(datasetId), dataJson);
    }

    String now = Instant.now().toString();
    DatasetMetadata updated = DatasetMetadata.builder().datasetId(datasetId).size(size)
        .schema(request.getSchema() != null ? request.getSchema() : existing.getSchema())
        .datasetType(existing.getDatasetType())
        .targetAction(
            request.getTargetAction() != null ? request.getTargetAction() : existing.getTargetAction())
        .metricRefs(request.getMetricRefs() != null ? request.getMetricRefs() : existing.getMetricRefs())
        .version(existing.getVersion() + 1).createTime(existing.getCreateTime()).updateTime(now).build();

    // Update the index
    indexCache.put(datasetId, updated);
    saveIndex(indexCache);

    logger.info("Updated dataset: {} (version {})", datasetId, updated.getVersion());
    return updated;
  }

  @Override
  public List<DatasetSample> getDataset(String datasetId) throws Exception {
    if (!indexCache.containsKey(datasetId)) {
      throw new IllegalArgumentException("Dataset not found: " + datasetId);
    }

    Path datasetFile = getDatasetFile(datasetId);
    if (!Files.exists(datasetFile)) {
      throw new IllegalArgumentException("Dataset data file not found: " + datasetId);
    }

    String content = Files.readString(datasetFile);
    return JsonUtils.getObjectMapper().readValue(content, new TypeReference<List<DatasetSample>>() {
    });
  }

  @Override
  public List<DatasetMetadata> listDatasets() throws Exception {
    // Reload index to get latest data
    loadIndex();
    return new ArrayList<>(indexCache.values());
  }

  @Override
  public void deleteDataset(String datasetId) throws Exception {
    if (!indexCache.containsKey(datasetId)) {
      throw new IllegalArgumentException("Dataset not found: " + datasetId);
    }

    // Delete the data file
    Path datasetFile = getDatasetFile(datasetId);
    Files.deleteIfExists(datasetFile);

    // Update the index
    indexCache.remove(datasetId);
    saveIndex(indexCache);

    logger.info("Deleted dataset: {}", datasetId);
  }
}
